File: tools/bake_lighting/experiments/reference_matching/building_review_analysis.py
"""Separate diffuse-model and AO differences on the custom pose's opaque material masks."""
import json,sys
from pathlib import Path
import numpy as np
import OpenImageIO as oiio
sys.path.insert(0,str(Path(__file__).resolve().parent.parent/'lighting_configurations/postprocess'))
from color_pipeline import ExrPasses,three_aces,write_png,save_json,srgb_decode,Y

capture,reference,output=map(Path,sys.argv[1:4])
full=json.loads((reference/'full/renders.json').read_text())
control=json.loads((reference/'geometric_diffuse/renders.json').read_text())
pointer=json.loads((reference/'full/scene.json').read_text())
scene=json.loads(Path(pointer['manifest']).read_text())
source=json.loads(Path(scene['sourceManifest']).read_text())
result={'groundCoverage':source['groundCoverage'],'materials':[],
    'conditions':'Fixed exposure, ACESFilmic, no grade change. Primary geometric Lambert control is a diagnostic, not the physical target.',
    'limits':['Global environment reflections lack local specular occlusion/parallax.',
              'Game material AO is measured separately; diffuse control disables Cycles bump correction and primary glossy response.',
              'Material masks exclude glass but procedural hooks and secondary material transport can still differ.']}
if (capture/'custom_raw').is_dir():
    target=next(r for r in full if r['pose']=='pose_custom');lambert=next(r for r in control if r['pose']=='pose_custom')
    metadata=json.loads((capture/'custom_raw/metadata.json').read_text())
    shape=(metadata['height'],metadata['width'],4);exposure=2**target['exposureEv']
    if abs(exposure-metadata['exposure'])>1e-8:raise RuntimeError('Exposure mismatch')
    def game(name):return np.flipud(np.fromfile(capture/'custom_raw'/(name+'.rgba32f'),dtype='<f4').reshape(shape))[:,:,:3]
    def diffuse(exr):return (exr.read('Diffuse Direct')+exr.read('Diffuse Indirect'))*exr.read('Diffuse Color')
    physical=ExrPasses(target['file'],'pose_custom');geometric=ExrPasses(lambert['file'],'pose_custom')
    images={'game_diffuse':game('diffuse'),'game_diffuse_no_ao':game('diffuse_no_material_ao'),
            'cycles_diffuse':diffuse(physical),'geometric_diffuse':diffuse(geometric)}
    for name,image in images.items():
        if image.shape!=shape[:2]+(3,) or not np.isfinite(image).all():raise RuntimeError('Invalid radiance')
        write_png(output/(name+'.png'),three_aces(np.maximum(image,0),exposure))
    def screenshot(file):
        image=oiio.ImageInput.open(str(file))
        if not image:raise RuntimeError('Missing comparison image: '+str(file))
        try:rgb=image.read_image(format=oiio.FLOAT)[:,:,:3]
        finally:image.close()
        if rgb.shape!=shape[:2]+(3,) or not np.isfinite(rgb).all():raise RuntimeError('Invalid comparison image')
        return rgb
    beauty={'original':screenshot(capture/'pose_custom-original.png'),
            'reflections':screenshot(capture/'pose_custom-reflections.png'),
            'cycles':screenshot(target['image'])}
    result['beautyMetric']='Display-linear luminance and sRGB RGB mean absolute error over the same eroded opaque masks; not scene-linear irradiance.'
    indices=physical.read('Material Index',channels=['X'])[:,:,0]
    audit={a['id']:a for a in scene['materialAudit']}
    for value in np.unique(indices):
        if value<=0:continue
        name=scene['build']['materialMasks'][str(int(value))];material=audit.get(name,{})
        if material.get('scope')!='city' or material.get('role')!='authored' or material.get('interiorProxy') or material.get('metalness',0)>.2 or material.get('opacity',1)<1:continue
        mask=indices==value
        for axis in [0,1]:
            for shift in [-2,-1,1,2]:mask &= np.roll(indices,shift,axis)==value
        # Only the building facade, away from road, bus and lower contact edges.
        region=np.zeros(indices.shape,bool);region[:int(shape[0]*.48),int(shape[1]*.25):int(shape[1]*.97)]=True;mask &= region
        if mask.sum()<1000:continue
        means={key:float((image[mask]@Y).mean()) for key,image in images.items()}
        beauty_means={key:float((srgb_decode(image[mask])@Y).mean()) for key,image in beauty.items()}
        errors={key:float(np.abs(beauty[key][mask]-beauty['cycles'][mask]).mean()) for key in ['original','reflections']}
        result['materials'].append({'id':name,'pixels':int(mask.sum()),'meanLinearY':means,
            'gameToFullDiffuse':means['game_diffuse_no_ao']/max(means['cycles_diffuse'],1e-9),
            'gameToGeometricDiffuse':means['game_diffuse_no_ao']/max(means['geometric_diffuse'],1e-9),
            'beauty':{'meanDisplayLinearY':beauty_means,'meanAbsoluteSrgbError':errors,
                'errorReductionPercent':100*(1-errors['reflections']/max(errors['original'],1e-9))}})
    result['materials'].sort(key=lambda row:-row['pixels'])
    result['exposure']=exposure
save_json(output/'analysis.json',result)
print(json.dumps(result),flush=True)

File: tools/bake_lighting/experiments/reference_matching/BuildingReviewAnalysis.mjs
// Independently assemble pose-grouped images and raw-material diagnostics.
import path from 'node:path';
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { authenticated, receipt, resultFiles } from '../lighting_configurations/StageInputs.mjs';
import { withGameBrowser } from '../lighting_configurations/capture_baselines/GameBrowser.mjs';
import { listFiles } from '../../../baking/Files.mjs';
import { outputPath, TOOL } from './Baseline.mjs';

export async function buildingReviewAnalysis(ctx) {
    if(ctx.publish||!ctx.options.capture||!ctx.options.reference||!ctx.options.output)throw new Error('capture, reference, new output required; diagnostic only');
    const capture=outputPath(ctx.root,ctx.options.capture),reference=outputPath(ctx.root,ctx.options.reference),output=outputPath(ctx.root,ctx.options.output);
    await authenticated(path.join(capture,'capture_receipt.json'));await authenticated(path.join(reference,'building_reference_receipt.json'));
    await mkdir(output);
    const review=JSON.parse(await readFile(path.join(capture,'review.json'),'utf8'));
    const renders=JSON.parse(await readFile(path.join(reference,'full/renders.json'),'utf8'));
    const request=JSON.parse(await readFile(path.join(reference,'full/request.json'),'utf8'));
    if(JSON.stringify(review.poses)!==JSON.stringify(request.poses))throw new Error('Game/reference poses differ');
    for(const run of review.runs)if(Math.abs(run.evidence.renderer.exposure-2**request.exposureEv)>1e-8)throw new Error('Game/reference exposure differs');
    await ctx.process(ctx.config.pythonExecutable,[path.join(ctx.root,TOOL,'building_review_analysis.py'),capture,reference,output]);
    await withGameBrowser(ctx,{width:2880,height:586},async page=>{
        const sheet=async(name,panels)=>{
            const images=await Promise.all(panels.map(async([label,file])=>({label,src:'data:image/png;base64,'+(await readFile(file)).toString('base64')})));
            await page.setContent(`<html><style>*{box-sizing:border-box}body{margin:0;display:flex;background:#101921;color:#d8e0e7;font:23px Arial}section{width:960px}header{height:46px;padding:10px 14px}img{display:block;width:960px;height:540px}</style>${images.map(im=>`<section><header>${im.label}</header><img src="${im.src}"></section>`).join('')}</html>`);
            await page.evaluate(()=>Promise.all([...document.images].map(im=>im.decode())));
            await page.screenshot({path:path.join(output,name+'.png')});
        };
        for(const pose of review.poses){
            const render=renders.find(r=>r.pose===pose.id);if(!render)throw new Error('Missing Cycles pose');
            await sheet(pose.id,[['Game · original',path.join(capture,pose.id+'-original.png')],['Game · opaque reflections',path.join(capture,pose.id+'-reflections.png')],['Cycles · corrected road coverage',render.image]]);
        }
        if(review.poses.some(p=>p.id==='pose_custom'))await sheet('custom_diffuse',[['Game diffuse · material AO off',path.join(output,'game_diffuse_no_ao.png')],['Cycles · full diffuse',path.join(output,'cycles_diffuse.png')],['Cycles · primary geometric Lambert',path.join(output,'geometric_diffuse.png')]]);
    });
    const table=['| Pose | Reflections | Passes | GPU median ms | GPU p1–p99 ms | CPU median ms | Calls | Triangles | Tex / Geo / Programs |','|---|---|---:|---:|---:|---:|---:|---:|---|',
        ...review.summary.map(r=>`| ${r.pose} | ${r.variant} | ${r.passes} | ${r.gpuMs.median.toFixed(2)} | ${r.gpuMs.p01.toFixed(2)}–${r.gpuMs.p99.toFixed(2)} | ${r.cpuMs.median.toFixed(2)} | ${r.calls.median} | ${r.triangles.median} | ${r.textures.median} / ${r.geometries.median} / ${r.programs.median} |`)];
    const analysis=JSON.parse(await readFile(path.join(output,'analysis.json'),'utf8'));
    const materials=['| Custom-pose opaque material | Pixels | Reflection RGB error reduction | Game diffuse / Cycles full | Game diffuse / geometric control |',
        '|---|---:|---:|---:|---:|',...analysis.materials.map(row=>`| ${row.id} | ${row.pixels} | ${row.beauty.errorReductionPercent.toFixed(1)}% | ${row.gameToFullDiffuse.toFixed(3)} | ${row.gameToGeometricDiffuse.toFixed(3)} |`)];
    await writeFile(path.join(output,'comparison.md'),review.conditions+'\n\n'+table.join('\n')+'\n\n'+review.memoryPolicy
        +'\n\n'+materials.join('\n')+'\n\n'+(analysis.beautyMetric??'No custom-pose material analysis.')
        +' Diffuse ratios use scene-linear luminance with game material AO removed. Image-error reductions are local measurements, not a full-scene parity score.\n');
    const file=path.join(output,'analysis_receipt.json');return resultFiles(await receipt(file,ctx.key,{capture,reference,output,diagnosticOnly:true},await listFiles(output)),file);
}
